Feature/new css floats lesson

mgrigoriev8109 opened the pull request:
closes https://github.com/TheOdinProject/top-meta/issues/54 with a new lesson on floats

rachelfearon commented on html_css/floats.md:
```diff
@@ -0,0 +1,39 @@
+
+The way the `float` property works is that it takes an element out of the normal layout flow of the document, puts that element to the side, and lets the other content flow around it. So if our element is an image and we give it the property `float: left;`, it will now be taken out of the normal layout flow and placed in the left-hand side of its container. The remaining content, such as a paragr ...
+
+Let's say we want a second paragraph to *not* wrap around the image, but to be displayed not only beneath the first paragraph but also entirely beneath the image. We can then use `clear: left;` to allow this second paragraph to clear the image that has been floated left. `clear` can be assigned left, right, or both to clear items which have been floated to the left, right, or in both directions.
```
Now, let's say we want a second paragraph in the same container to *not* wrap around the image, but to be displayed beneath **both** the first paragraph and the floated image. We can then use `clear: left;` to allow this second paragraph to clear the image that has been floated left. `clear` can be assigned `left`, `right`, or `both` to clear items which have been floated to the left, right, or in both directions.

rachelfearon commented on html_css/floats.md:
```diff
@@ -0,0 +1,39 @@
+Let's say we want a second paragraph to *not* wrap around the image, but to be displayed not only beneath the first paragraph but also entirely beneath the image. We can then use `clear: left;` to allow this second paragraph to clear the image that has been floated left. `clear` can be assigned left, right, or both to clear items which have been floated to the left, right, or in both directions.
+
+### Problems with float
+While `float` was first used to wrap text around an image, as developers realised that you can float *anything* the property quickly became a commonplace tool for creating creating webpage layouts. This led to problems such as the one we're about to discuss'.
```
While `float` was first used to wrap text around an image, developers soon realized that you can float almost *anything*, and the property quickly became a commonplace tool for creating webpage layouts. However, this led to problems, such as the one we're about to discuss.

rachelfearon commented on html_css/floats.md:
```diff
@@ -0,0 +1,39 @@
+### Problems with float
+While `float` was first used to wrap text around an image, as developers realised that you can float *anything* the property quickly became a commonplace tool for creating creating webpage layouts. This led to problems such as the one we're about to discuss'.
+
+Say you want to take your image, and your first paragraph, and put them into a box that will contain those two elements. If we attempt to `float: left;` the image so that the first paragraph is on its right, and then `clear` the second paragraph past the box container - it won't work! This happens because the float was taken out of the normal flow of the document, and even though the second parag ...
```
Say you want to take your image and your first paragraph, and put them into a box that will contain those two elements. If we attempt to `float: left;` the image so that the first paragraph is on its right, and then `clear` a second paragraph past the box container - it won't work! This happens because the float was taken out of the normal flow of the document, and even though the second paragraph comes after the first paragraph, it will no longer clear. The most modern solution to this problem is to use `display: flow-root`, which creates a miniature layout within the page, inside of which everything is contained.

rachelfearon commented on html_css/floats.md:
```diff
@@ -0,0 +1,39 @@
+
+Say you want to take your image, and your first paragraph, and put them into a box that will contain those two elements. If we attempt to `float: left;` the image so that the first paragraph is on its right, and then `clear` the second paragraph past the box container - it won't work! This happens because the float was taken out of the normal flow of the document, and even though the second parag ...
+
+Problems such as this one used to require various hacks to work around, and are part of the reason why using floats to create layouts is now considered a [legacy layout method](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Legacy_Layout_Methods), and using CSS Grid or Flexbox is preferred.
```
Problems such as this used to require various hacks to work around, and are part of the reason why using floats to create layouts is now considered a [legacy layout method](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Legacy_Layout_Methods). Today, using CSS Grid or Flexbox is preferred.

rachelfearon commented on html_css/floats.md:
```diff
@@ -0,0 +1,39 @@
+
+### Additional Resources
+This section contains helpful links to other content. It isn't required, so consider it supplemental for if you need to dive deeper into something
+* [For a more in-depth look at float](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Floats) MDN has some great documentation on `float`.
```
* [For a more in-depth look at float](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Floats) MDN has some great documentation.

rachelfearon commented on html_css/floats.md:
```diff
@@ -0,0 +1,39 @@
+
+### Learning Outcomes
+* You'll learn how to properly use floats. 
+* You'll learn why to avoid using floats the way they were used in the past, and their history.
```
* Learn how floats work and how to use them
* Understand how floats were used in the past, and why to avoid these methods in the present

digidub commented on html_css/floats.md:
```diff
@@ -0,0 +1,39 @@
+* You'll learn why to avoid using floats the way they were used in the past, and their history.
+
+### Present Day Use of Floats
+In the past `float` was commonly used to create entire webpage layouts. While we will discuss this later on in the lesson, it is important to know that in the present day `float` should **only** be used to wrap text around images. When working with layout, stick with Flexbox or Grid.
```
```suggestion
In the past, `float` was commonly used to create entire webpage layouts. While we will discuss this later on in the lesson, it is important to know that in the present day `float` should **only** be used to wrap text around images. When working with layouts, stick with Flexbox or Grid.
```

codyloyd commented:
Content here is good.. nice work!

I think we should frame the topic a little more positively though.  I think the lesson hits the whole "floats have been used for hacky layout stuff" message a little hard from the beginning.

I'd rather remove stuff like that from the introduction and middle sections, focus on the correct usage and pretend that the hacky layout stuff doesn't even exist until that section at the end.  ALSO: i'd move that section to the very end (after "shapes").   This change should probably also effect the headings... like `present day usage of floats` should probably be something like `basic usage`.

AAAAAAND finally... I think it would be great to get a couple of examples in here with a codepen embed.

see this lesson for an example and/or reach out to me on discord for assistance. https://github.com/TheOdinProject/curriculum/blob/main/foundations/html_css/html-foundations/working-with-text.md

basically you create the codepen, click the 'embed' button at the bottom and copy/paste the provided code into the lesson.

We have an Odin-owned codepen for examples, but if you want to go ahead and create them on your personal account, we can fork it from the TOP one and switch the embed later.

mgrigoriev8109 commented:
Completely see your point Cody, changed the structure and added in the Codepens. 

Let me know if anyone else sees anything else this lesson could improve on.

mgrigoriev8109 commented:
> Content here is good.. nice work!
> 
> I think we should frame the topic a little more positively though. I think the lesson hits the whole "floats have been used for hacky layout stuff" message a little hard from the beginning.
> 
> I'd rather remove stuff like that from the introduction and middle sections, focus on the correct usage and pretend that the hacky layout stuff doesn't even exist until that section at the end. ALSO: i'd move that section to the very end (after "shapes"). This change should probably also effect the headings... like `present day usage of floats` should probably be something like `basic usage`.
> 
> AAAAAAND finally... I think it would be great to get a couple of examples in here with a codepen embed.
> 
> see this lesson for an example and/or reach out to me on discord for assistance. https://github.com/TheOdinProject/curriculum/blob/main/foundations/html_css/html-foundations/working-with-text.md
> 
> basically you create the codepen, click the 'embed' button at the bottom and copy/paste the provided code into the lesson.
> 
> We have an Odin-owned codepen for examples, but if you want to go ahead and create them on your personal account, we can fork it from the TOP one and switch the embed later.

Completely see your point Cody, changed the structure and added in the Codepens.

Let me know if anyone else sees anything else this lesson could improve on.

codyloyd commented on html_css/floats.md:
```diff
@@ -0,0 +1,59 @@
+</p>
+<script async src="https://cpwebassets.codepen.io/assets/embed/ei.js"></script>
+
+Now talk about a problem we might encounter using `float`, `clear`, and additional containers. Say you want to take your box and your first paragraph, and put them into a another box that will contain those two elements. If we attempt to `float: left;` the box so that the first paragraph is on its right, and then `clear` a second paragraph past the box container, you'll see below that the second  ...
```
`Now talk about` should probably be `Now let's talk about`

or something :D

codyloyd reviewed (changes requested):
one small issue :D  good fixes all around

mgrigoriev8109 reviewed (commented):
Thanks for catching that, I just fixed the syntax.
